crates/wasi/src/preview2/ctx.rs

```diff
@@ -258,3 +258,53 @@ pub struct WasiCtx {
     pub(crate) stdout: u32,
     pub(crate) stderr: u32,
 }
+
+impl WasiCtx {
+    /// Flush all buffered output and wait for it to reach the destination.
+    ///
+    /// NOTE: This function should be used when [`WasiCtx`] is used in an async embedding
+    /// (i.e. with [`crate::preview2::command::add_to_linker`]). Use its counterpart
+    /// `sync_join_background_tasks` in a synchronous embedding (i.e. with
+    /// [`crate::preview2::command::sync::add_to_linker`].
+    ///
+    /// In order to implement non-blocking streams, we often often need to offload async
+    /// operations to background `tokio::task`s. These tasks are aborted when the resources
+    /// in the `Table` referencing them are dropped. In some cases, this abort may occur before
+    /// buffered output has been flushed. Use this function to wait for all
+    /// written data to reach its destination gracefully, even if wasm didn't
+    /// explicitly wait for this.
+    ///
+    /// In some embeddings, a misbehaving client might cause this graceful exit to await for an
+    /// unbounded amount of time, so we recommend bounding this with a timeout or other mechanism.
```

Comment:
I do wonder if this should also take a timeout just like the sync version. This might encourage users to do the right thing and it has the added benefit of being symetrical with the sync version.